### ISSUE TITLE
Tracking of 5.1 update seems unreliable

### ISSUE THREAD

**Issue by POEDaley** [cboulay/PSMoveService#232]
After calibrating properly in 5.0, my controllers worked fine with very little need to adjust.  Using this with Riftcat, I only had to recenter riftcat on very rare occasions.

Since 5.2, and the need to press the controllers to your face and hit the 2 buttons, I've had really bad controller drift.  They would end up far out of place within a couple minutes of playing, and no matter how often I reset using the buttons, the controllers would always be pretty out of wack.  Once they drifted, only restarting PS move service would correct it, albeit for a short period of time again.

A few things that would happen when using the 2 button hotkey:
- Controllers would center properly again, but would have a bad rotation on one or 2 of the controllers
- Controllers would not center properly
- Controllers would center properly, but move at a 90 degree angle

I switched back to 5.0 to see if it was perhaps my 2 year old who dropped the controllers too much, but once calibrated again in 5.0, it worked perfectly again, and for long periods of time.  I really want to go back up to 5.2 for the 5% drop in CPU performance, but it's unplayable for me due to the drift.

Is this normal, or am I missing some configuration steps in 5.1?

Thanks,

**Comment by HipsterSloth** [comment]
-When you say "drift" you mean position drift or or orientation drift? (or both?). 
-After you notice the drift, if you lauch the "test tracking" tool in PSMoveConfig is that off as well? Or is it only off in steam?

**Comment by POEDaley** [comment]
The answer is "both" on both accounts.

While calibrating the controllers, if I calibrated them in sequence, by the time I finished the 3rd controller and came back, it was off by about 60-70 degrees in rotation.  I checked specifically by laying the controller 0 down on a surface pointing directly toward the front at 0 degrees along the green line.  Without touching it, it would stray on its own.

In game, it would demonstrate the same thing.  It would also stray about the same degree in direction.  So if I held the controller horizontal pointing directly forwards, and extended forwards, it would move forwards in the direction it was pointing in game along that same offset angle.

I'd be happy to do any further testing you want, but I just have a question how to maintain my current settings and calibration of the 5.0 setup I'm running.  Just to be certain, am I able to just rename my appdata folder and run the 5.2 download in a seperate folder?  This way I just need to name it back correctly in appdata to restore my previous setup?  Or is there something else that I would need to change as well?

Thanks for your attention on this.  I can't stress enough how much I appreciate your work.

**Comment by HipsterSloth** [comment]
Interesting. It's possible the orientation drift is affecting the ability to align the tracking spaces correctly. I just did a test locally on the new 5.3 release (https://github.com/cboulay/PSMoveService/releases/tag/v0.9-alpha5.3) and I'm not seeing the same issue. Not that I think that's fixed, just that I think the drift is dependent on local conditions. I think I need to expose a setting for the magnetometer that effects how much we attempt to blend in the magnetometer reading. I think for some people it's too low, which leads to yaw drift.

> Just to be certain, am I able to just rename my appdata folder and run the 5.2 download in a seperate folder? This way I just need to name it back correctly in appdata to restore my previous setup? Or is there something else that I would need to change as well?

Yup that should be fine. All settings are stored in %appdata%\PSMoveService so backing up and restoring that folder between version is fine. Also, always make sure to re-run `SteamVR_ReinstallDriverWin64.bat` when switching back and forth between release (the client API used by the steamvr plugin can vary per release as new features are added).

**Comment by POEDaley** [comment]
Quick update - 

This issue may have been partly user error.  I tested this again using 5.3 this morning and my first controller calibrated without yaw drift, then the next 2 failed and had the same results as above.  I recalibrated those 2 controllers a few more times, and restarted the service after each calibration and finally had all 3 controllers stable.

Controller 0 - 1 attempt to get stable
Controller 1 - 3 attempts to get stable
Controller 2 - 5 attempts to get stable

I don't think 5.2 - 5.3 fixed the issue since I had such trouble getting it done right. But I'm happy to report it seems to be working correctly in the config tool.

I will test this out with real use later tonight, but I wanted to update you that I did manage to get it right, so it's not broken.  It may be just a bit finicky, but correctable with some perseverance.  

If there is anything else I can do or test, let me know!

Thanks,

**Comment by HipsterSloth** [comment]
> I don't think 5.2 - 5.3 fixed the issue since I had such trouble getting it done right. 

Agreed. Sounds like there are some improvements that could be made to the magnetometer calibration. Would you mind attaching the %appdata%\PSMoveService\xx_xx_xx_xx_xx_xx.json config file for a controller with a config attempt that didn't work and then a second version for when it does. I'm curious if there is a notable difference in the magnetometer calibration values between attempts.

In case your curious the magnetometer values in question look like this:

```
        "Magnetometer": {
            "Center": {
                "X": "65",
                "Y": "-121",
                "Z": "92"
            },
            "BasisX": {
                "X": "1",
                "Y": "0",
                "Z": "0"
            },
            "BasisY": {
                "X": "0",
                "Y": "1",
                "Z": "0"
            },
            "BasisZ": {
                "X": "0",
                "Y": "0",
                "Z": "1"
            },
            "Extents": {
                "X": "146",
                "Y": "146",
                "Z": "140"
            },
```

Where the "center" is the midpoint of the bounding box that encloses the magnetometer readings, the extents are the half-widths of the bounding box (should be close to the same value across each axis), and the "identity" is the direction of magnetic north recorded in the second step of the magnetometer calibration (i.e. when your controller is aligned with the axes of the calibration mat).

**Comment by HipsterSloth** [comment]
I'm updating the bug title to be about fixing the controller drift issue specifically.

**Comment by POEDaley** [comment]
I ended up getting lucky and first calibration went poorly with some mild, but consistant yaw, and second calibration went well with a stable controller.  Here are the 2 calibrations of the same controller.

[Controller Calibrations.zip](https://github.com/cboulay/PSMoveService/files/520960/Controller.Calibrations.zip)

**Comment by HipsterSloth** [comment]
Thanks for taking the time to record those! It looks like the biggest difference is in the magnetometer extents. In the good config file:

```
            "Extents": {
                "X": "168",
                "Y": "171",
                "Z": "170.5"
            },
```

And in the bad config file:

```
                "X": "168.5",
                "Y": "175",
                "Z": "155"
```

Note in the good case that the extents are about the same. This is typically what they should be in an environment with little to no magnetic distortions. In the bad calibration (which isn't that bad) the z-axis is a bit smaller than the true extents. This distorts the magnetometer readings a bit from their true direction, which will lead to drift.

In the mean time, you can prevent drift by making sure you're getting even coverage of the magnetometer sample points (try and maximize the extents). You can tell you have this if the extent values are within around +/- 5 of each other. I have the extent numbers circled in this diagram (from another thread):

![magnetometer](https://cloud.githubusercontent.com/assets/16793853/19258925/974d9b7c-8f31-11e6-9c18-317d9d5e1ad9.jpg)

Now all that said, it sucks that this calibration is that finicky. This can actually be fixed by computing a best fit ellipsoid of the sample points to get the true magnetometer readings shape. I have one algorithm I wrote a while for this, this "min volume fitting", but that turned out to be the non ideal fit. I need to instead implement a "least squared error fit". Then this wont be so sensitive. I'm going to update the title of the bug to reflect this issue.

**Comment by POEDaley** [comment]
That helps a lot, thanks!  What's happening makes a lot of sense now.  

I knew when I calibrated that it wasn't all that horrible, but it did continue to drift so I figured it would show.

Can any of those values be manually adjusted in that json file?  For example, could we tweak the extents say from 168-170 to tighten it up?  Also, and more importantly, are you able to adjust the yaw angle of the controller in there?  This way I can ensure that all 3 controllers are lined up 100%.  Or would this just cause issues down the road?

In closing, I would consider this issue resolved for myself now.  But the work your doing and this software absolutely fascinates me! lol

**Comment by HipsterSloth** [comment]
Yeah you can manually tweak values in there (the config file is the authority on the controller settings). 

You can tweak the yaw, but it's not intuitive. It's effectively stored as the direction of the magnetic field that was recorded when the controller was aligned with the calibration mat as the "indentity" entry. For example here is mine:

```
"X": "-0.389402121",
"Y": "0.914950252",
"Z": "-0.105981141"
```

You can manually compute this direction if you know the yaw and pitch of the identity direction:

```
x = cos(yaw)*cos(pitch)
y = sin(pitch)
z = sin(yaw)*cos(pitch)
```

Conversely if you want to get the yaw and pitch of your current magnetometer direction:

```
pitch = asin(y)
yaw = acos(x / cos(pitch))
```

It would be pretty easy to add a function to the controller config panel to let you tweak the yaw so that you don't have to do this manual trig.
